server/src/main/java/org/opensearch/common/cache/stats/StatsHolder.java

```diff
@@ -0,0 +1,161 @@
+        this.dimensionNames = dimensionNames;
+        this.statsMap = new ConcurrentHashMap<>();
+    }
+
+    public List<String> getDimensionNames() {
+        return dimensionNames;
+    }
+
+    public ConcurrentMap<Key, CacheStatsResponse> getStatsMap() {
+        return statsMap;
+    }
+
+    // For all these increment functions, the dimensions list comes from the key, and contains all dimensions present in dimensionNames.
+    // The order doesn't have to match the order given in dimensionNames.
+    public void incrementHits(ICacheKey<?> key) {
+        internalIncrement(key.dimensions, (response, amount) -> response.hits.inc(amount), 1);
+    }
+
+    public void incrementMisses(ICacheKey<?> key) {
+        internalIncrement(key.dimensions, (response, amount) -> response.misses.inc(amount), 1);
+    }
+
+    public void incrementEvictions(ICacheKey<?> key) {
+        internalIncrement(key.dimensions, (response, amount) -> response.evictions.inc(amount), 1);
+    }
+
+    public void incrementSizeInBytes(ICacheKey<?> key, long amountBytes) {
+        internalIncrement(key.dimensions, (response, amount) -> response.sizeInBytes.inc(amount), amountBytes);
+    }
+
+    public void incrementEntries(ICacheKey<?> key) {
+        internalIncrement(key.dimensions, (response, amount) -> response.entries.inc(amount), 1);
+    }
+
+    public void decrementEntries(ICacheKey<?> key) {
+        internalIncrement(key.dimensions, (response, amount) -> response.entries.inc(amount), -1);
+    }
+
+    /**
+     * Reset number of entries and memory size when all keys leave the cache, but don't reset hit/miss/eviction numbers
+     */
+    public void reset() {
+        for (Key key : statsMap.keySet()) {
+            CacheStatsResponse response = statsMap.get(key);
+            response.sizeInBytes.dec(response.getSizeInBytes());
+            response.entries.dec(response.getEntries());
+        }
+    }
+
+    public long count() {
+        // Include this here so caches don't have to create an entire CacheStats object to run count().
+        long count = 0L;
+        for (Map.Entry<Key, CacheStatsResponse> entry : statsMap.entrySet()) {
+            count += entry.getValue().getEntries();
+        }
+        return count;
+    }
+
+    private void internalIncrement(List<CacheStatsDimension> dimensions, BiConsumer<CacheStatsResponse, Long> incrementer, long amount) {
+        assert dimensions.size() == dimensionNames.size();
+        CacheStatsResponse stats = internalGetStats(dimensions);
+        incrementer.accept(stats, amount);
+    }
+
+    private CacheStatsResponse internalGetStats(List<CacheStatsDimension> dimensions) {
+        Key key = new Key(getOrderedDimensionValues(dimensions, dimensionNames));
+        CacheStatsResponse response = statsMap.get(key);
+        if (response == null) {
+            response = new CacheStatsResponse();
+            statsMap.put(key, response);
+        }
```

Comment:
Use `computeIfAbsent`. Never `get` then `put` on a concurrent data structure. You're asking for a race condition.